From c0bd5dd660766f58887b00d8be0137bc042e0614 Mon Sep 17 00:00:00 2001
From: Ecco park <eccopark@broadcom.com>
Date: Wed, 3 Dec 2014 17:22:06 -0800
Subject: [PATCH 25/27] net: wirelss: bcmdhd fix roaming bug

1) Bug : 18541286 wifi stuck on connected to an AP when N6 was miles away
we should update BSSID in wl_bss_roaming_done, not wl_notify_status for roaming case
It causes out-of-sync between FW and DHD
Signed-off-by: Ecco park <eccopark@broadcom.com>

Change-Id: I75e706d63aa962c8c2640797e2af6811990c989d
---
 drivers/net/wireless/bcmdhd/dhd_pno.c     | 22 ++++++++++++++++++----
 drivers/net/wireless/bcmdhd/wl_cfg80211.c | 10 ++++++----
 2 files changed, 24 insertions(+), 8 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/dhd_pno.c b/drivers/net/wireless/bcmdhd/dhd_pno.c
index 515e672..cd21965 100644
--- a/drivers/net/wireless/bcmdhd/dhd_pno.c
+++ b/drivers/net/wireless/bcmdhd/dhd_pno.c
@@ -1285,8 +1285,15 @@ exit:
 		_dhd_pno_reinitialize_prof(dhd, _params, DHD_PNO_LEGACY_MODE);
 exit_no_clear:
 	/* clear mode in case of error */
-	if (err < 0)
-		_pno_state->pno_mode &= ~DHD_PNO_LEGACY_MODE;
+	if (err < 0) {
+		int ret = dhd_pno_clean(dhd);
+		if (ret < 0) {
+			 DHD_ERROR(("%s : failed to call dhd_pno_clean (err: %d)\n",
+			 	__FUNCTION__, ret));
+		} else {
+			_pno_state->pno_mode &= ~DHD_PNO_LEGACY_MODE;
+		}
+	}
 	return err;
 }
 int
@@ -2095,8 +2102,15 @@ dhd_pno_set_for_gscan(dhd_pub_t *dhd, struct dhd_pno_gscan_params *gscan_params)
 
 exit:
 	/* clear mode in case of error */
-	if (err < 0)
-		_pno_state->pno_mode &= ~DHD_PNO_GSCAN_MODE;
+	if (err < 0) {
+		int ret = dhd_pno_clean(dhd);
+		if (ret < 0) {
+			 DHD_ERROR(("%s : failed to call dhd_pno_clean (err: %d)\n",
+			 	__FUNCTION__, ret));
+		} else {
+			_pno_state->pno_mode &= ~DHD_PNO_GSCAN_MODE;
+		}
+	}
 	kfree(pssid_list);
 	kfree(p_pfn_significant_bssid);
 	kfree(p_pfn_bssid);
diff --git a/drivers/net/wireless/bcmdhd/wl_cfg80211.c b/drivers/net/wireless/bcmdhd/wl_cfg80211.c
index ba9118c..e43d0ac 100644
--- a/drivers/net/wireless/bcmdhd/wl_cfg80211.c
+++ b/drivers/net/wireless/bcmdhd/wl_cfg80211.c
@@ -8428,14 +8428,16 @@ wl_notify_connect_status(struct bcm_cfg80211 *cfg, bcm_struct_cfgdev *cfgdev,
 			if (!wl_get_drv_status(cfg, DISCONNECTING, ndev)) {
 					printk("wl_bss_connect_done succeeded with " MACDBG "\n",
 						MAC2STRDBG((u8*)(&e->addr)));
+					/* In case of roaming, the update will be handled in roaming handler */
+					if (!wl_get_drv_status(cfg, CONNECTED, ndev)) {
+						wl_update_prof(cfg, ndev, e, &act, WL_PROF_ACT);
+						wl_update_prof(cfg, ndev, NULL, (void *)&e->addr, WL_PROF_BSSID);
+					}
 					wl_bss_connect_done(cfg, ndev, e, data, true);
 					WL_DBG(("joined in BSS network \"%s\"\n",
 					((struct wlc_ssid *)
 					 wl_read_prof(cfg, ndev, WL_PROF_SSID))->SSID));
-				}
-			wl_update_prof(cfg, ndev, e, &act, WL_PROF_ACT);
-			wl_update_prof(cfg, ndev, NULL, (void *)&e->addr, WL_PROF_BSSID);
-
+			}
 		} else if (wl_is_linkdown(cfg, e)) {
 			if (cfg->scan_request)
 				wl_notify_escan_complete(cfg, ndev, true, true);
-- 
1.9.1

